source/src/texk/kpathsea/kpsewhich.man
.TH KPSEWHICH 1 "4 January 1998" "Kpathsea @VERSION@"
.\"=====================================================================
.if n .ds MP MetaPost
.if t .ds MP MetaPost
.if n .ds MF Metafont
.if t .ds MF M\s-2ETAFONT\s0
.if t .ds TX \fRT\\h'-0.1667m'\\v'0.20v'E\\v'-0.20v'\\h'-0.125m'X\fP
.if n .ds TX TeX
.ie t .ds OX \fIT\v'+0.25m'E\v'-0.25m'X\fP\" for troff
.el .ds OX TeX\" for nroff
.\" the same but obliqued
.\" BX definition must follow TX so BX can use TX
.if t .ds BX \fRB\s-2IB\s0\fP\*(TX
.if n .ds BX BibTeX
.\" LX definition must follow TX so LX can use TX
.if t .ds LX \fRL\\h'-0.36m'\\v'-0.15v'\s-2A\s0\\h'-0.15m'\\v'0.15v'\fP\*(TX
.if n .ds LX LaTeX
.\"=====================================================================
.SH NAME
kpsewhich \- standalone path lookup and and expansion for kpathsea
.SH SYNOPSIS
.B kpsewhich
.RI [ options ]
.RI [ filenames ]
.\"=====================================================================
.SH DESCRIPTION
This manual page is not meant to be exhaustive.  The complete
documentation for this version of \*(TX can be found in the info file
or manual
.IR "Kpathsea: A library for path searching" .
.PP
.B kpsewhich
is used as a standalone front-end of the kpathsea library that can be
used to examine variables and find files.  When the
.B -format
option is not given, the search path used when looking for a file is
inferred from the name given, by looking for a known extension.  If
no known extension is found, the search path for \*(TX source files is
used.
.\"=====================================================================
.SH OPTIONS
.B kpsewhich
accepts the following options:
.TP
.BI -debug \ num
.rb
Set debugging flags.
.TP
.BI -D \ num
.rb
Use a base resolution of
.IR num ;
the default, set by the installer, is typically 600.
.TP
.BI -dpi \ num
As
.BR -D.
.TP
.BI -expand-braces \ string
.rb
Print variable and brace expansion of
.IR string .
.TP
.BI -expand-path \ string
.rb
Print complete path expansion of
.IR string .
.TP
.BI -expand-var \ string
.rb
Print variable expansion of
.IR string .
.TP
.BI -format \ name
.rb
Use file type
.IR name .
See the info manual for a list of valid names, or use the
.B -help
option to print the list.
.TP
.B -help
.rb
Print help message and exit.
.TP
.B -interactive
.rb
Ask for additional filenames to look up.
.TP
.BI -mktex \ fmt
.rb
enable
.RI mktex fmt
generation.
.RI ( fmt =pk/mf/tex/tfm)
.TP
.BI -mode \ string
.rb
Set device name for
.I $MAKETEX_MODE
to
.IR string ;
no default.
.TP
.B -must-exist
.rb
Search the disk as well as ls-R if necessary.
.TP
.BI -no-mktex \ fmt
.rb
disable
.RI mktex fmt
generation.
.RI ( fmt =pk/mf/tex/tfm)
.TP
.BI -path \ string
Search in the path
.IR string .
.TP
.BI -progname \ string
.rb
Set program name to
.IR string .
.\" .TP
.\" .BI -separator \ string
.\" .rb
.\" .I string
.\" separates components in
.\" .B -path
.\" output; default is
.\" .I :
.\" on UNIX systems.
.TP
.BI -show-path \ name
.rb
Output search path for file type
.IR name .
See the info manual for a list of valid names, or use the
.B -help
option to print the list.
.TP
.BI -var-value \ variable
.rb
Print the expansion of
.IR variable .
.B -version
.rb
Print version information and exit.
.\"=====================================================================
.SH "SEE ALSO"
.BR mktexlsr (1),
.BR mktexmf (1),
.BR mktexpk (1),
.BR mktextfm (1).


www/NEWS
-------------------------------------------------
pdfTeX 3.141592-1.40.7 was released on 2008-02-17
-------------------------------------------------
- Bugfix:
  - PDFs > 2GiB (LFS) were still broken on 32-bit systems

-------------------------------------------------
pdfTeX 3.141592-1.40.6 was released on 2008-02-04
-------------------------------------------------
- Bugfixes:
  - inclusion of encrypted PDFs was broken (861)
  - PDFs > 2GiB (LFS) would be broken, even on 64-bit systems. Now the basic
    support for LFS is there, you only need to supply the right compiler flags
    (this is not yet done automatically). See build.sh for hints on compiling.
  - Page groups in pdfs (see the discussion in the notes vor v1.10a) are now
    handled correctly (741)
  - When including transparent PNGs pdftex now automatically adds a Page Group
    (896)
  - broken map files would lead to core dumps
  - unusual TrueType fonts would lead to an assertion error (874)
  - the null object in the object table was never initialized
  - included PDFs with /PieceInfo lost their /LastModified date
  - an empty colorstack crashed pdftex (897)
- pdfTeX uses libpng 1.2.24
- pdfTeX uses xpdf 3.02pl2

-------------------------------------------------
pdfTeX 3.141592-1.40.5 was released on 2007-07-31
-------------------------------------------------
- Bugfixes: 
  - calling 'pdftex %' would crash (830)
  - fflush the recorder file
- pdfTeX uses xpdf 3.02pl1

-------------------------------------------------
pdfTeX 3.141592-1.40.4 was released on 2007-07-11
-------------------------------------------------
- Bugfixes: 
  - It was impossible to set the font Flags to 4 in map entries (762)
  - The font Flags of non-embedded fonts were wrong (794)
    Now pdfTeX uses the Flags from map entries (if found) or guesses values for
    the base14 fonts. It also warns when you not-embedd a font without Flags
  - Subsetted fonts still had their private UniqueID copied (774)
  - \pdfunescapehex returned a wrong result if the last character was not a
    hexadecimal digit and the last hex pair was incomplete (777)
  - Copying a box with \pdfstartlink inside produced the same /Link annotation
    object multiply, with same object number and identical coordinates on the
    page (799)
  - When trying to use a font without TFM, without map entry, and no PK font
    available, a broken PDF was produced. Now no PDF is generated and the error
    messages have been improved (721)
  - Problems with vf fonts (722)
  - With pdf inclusion the font descriptor could get mixed up and miss some
    required keys (810)
  - Font expansion led to wrong log output of \kerns (736)
  - \pdfprimitive was broken (see
    http://www.ntg.nl/pipermail/ntg-pdftex/2007-April/002860.html)
  - The -recorder option of web2c didn't record fonts, included images and some
    other files (772)
  - Using \pdfximage in IniTeX before \dump caused a crash when the format was
    used (824)
  - \letterspacefont had some problems (712)
  - \letterspacefont and \pdffontexpand had some problems (747)
- Improvements: 
  - pdf compression is now a bit faster (patch 726)
  - the behaviour of TeX when it asks for a file name has been improved (756)
- pdfTeX uses libpng 1.2.18
- pdfTeX uses xpdf 3.02

-------------------------------------------------
pdfTeX 3.141592-1.40.3 was released on 2007-02-04
-------------------------------------------------
- bugfix: document which used type 1 and truetype or opentype fonts resulted in
  broken pdfs (729)

-------------------------------------------------
pdfTeX 3.141592-1.40.2 was released on 2007-01-31
-------------------------------------------------
- bugfix: maplines starting with = would not work as advertised
- bugfix: xpdf would complain that PDF 1.7 is too new

-------------------------------------------------
pdfTeX 3.141592-1.40.1 was released on 2007-01-08
-------------------------------------------------
- bugfix: the shell escape function was a bit broken
- bugfix: \pdffontattr gave confusing warnings
- allow JBIG2 images only with PDF >= 1.4
- allow \pdfglyphtounicode in dvi mode for format generation
- pdfTeX uses libpng 1.2.15

-------------------------------------------------
pdfTeX 3.141592-1.40.0 was released on 2007-01-01
-------------------------------------------------
- pdfetex (as a separate binary/program) is gone; all extensions are now in
  pdftex. The make files still know about the target pdfetex, but it's just a
  copy of pdftex
- The sources of TeX, eTeX and pdfTeX have been merged into one file pdftex.web
  and one change file pdftex.ch. The sources of TeX and eTeX distributed are
  just there for reference
- New features:
  - pdfTeX can now generate PDF 1.5 object streams, which leads to smaller
    PDFs. This is controlled by setting \pdfobjcompresslevel to a value >0 (3
    compress everything, 2 don't compress /Info, 1 also don't compress embedded
    PDFs) and works only with PDF >=1.5
  - pdfTeX now supports another image file format: JBIG2 files (.jbig2)
  - the memory allocated for objects and destination names now grows
    dynamically
  - pdfTeX now generates smaller PDFs since for char positioning the width of
    chars (/Widths) is stored with more precision and so the chars must rarely
    be positioned separately
  - the PK resolution is now taken from the "pk_dpi" parameter in texmf.cnf if
    it has not been set in the format file or by the user
  - pdfTeX now removes the generated pdf in case of a fatal error
  - pdfTeX now sets /ModDate and /Trapped in the InfoDict by default (the
    values can be overridden with \pdfinfo)
  - the format of warnings and errors issued by pdfTeX has been changed and
    unified
  - the output of -version has been extended and now includes information about
    the libraries actually used, which can be handy when using shared libraries
  - rules and their positioning on the page are now inside a qQ group
- New primitives:

source/src/texk/kpathsea/kpsewhich.c
/* kpsewhich -- standalone path lookup and variable expansion for Kpathsea.
   Ideas from Thomas Esser and Pierre MacKay.

   Copyright (C) 1995 - 2005 Karl Berry & Olaf Weber.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

*/

#include <kpathsea/config.h>
#include <kpathsea/c-ctype.h>
#include <kpathsea/c-pathch.h>
#include <kpathsea/expand.h>
#include <kpathsea/getopt.h>
#include <kpathsea/line.h>
#include <kpathsea/pathsearch.h>
#include <kpathsea/proginit.h>
#include <kpathsea/str-list.h>
#include <kpathsea/tex-file.h>
#include <kpathsea/tex-glyph.h>
#include <kpathsea/variable.h>
#include <kpathsea/progname.h>


/* Base resolution. (-D, -dpi) */
unsigned dpi = 600;

/* For variable and path expansion.  (-expand-var, -expand-path,
   -show-path, -separator) */
string var_to_expand = NULL;
string braces_to_expand = NULL;
string path_to_expand = NULL;
string path_to_show = NULL;
string var_to_value = NULL;

/* Only match files in given subdirs.  (-subdir) */
str_list_type subdir_paths;

/* The file type and path for lookups.  (-format, -path) */
kpse_file_format_type user_format = kpse_last_format;
string user_format_string;
string user_path;

/* Interactively ask for names to look up?  (-interactive) */
boolean interactive = false;

/* Search the disk as well as ls-R?  (-must-exist) */
boolean must_exist = false;

/* Return all matches, not just the first one?  (-all) */
boolean show_all = false;

/* The device name, for $MAKETEX_MODE.  (-mode) */
string mode = NULL;

/* The program name, for `.PROG' construct in texmf.cnf.  (-program) */
string progname = NULL;

/* The engine name, for '$engine' construct in texmf.cnf.  (-engine) */
string engine = NULL;

/* Return the <number> substring in `<name>.<number><stuff>', if S has
   that form.  If it doesn't, return 0.  */

static unsigned
find_dpi P1C(string, s)
{
  unsigned dpi_number = 0;
  string extension = find_suffix (s);
  
  if (extension != NULL)
    sscanf (extension, "%u", &dpi_number);

  return dpi_number;
}

/* Use the file type from -format if that was specified, else guess
   dynamically from NAME.  Return kpse_last_format if undeterminable.
   This function is also used to parse the -format string, a case which
   we distinguish by setting is_filename to false.

   Note that a few filenames have been hard-coded for format types that
   differ from what would be inferred from their extensions. */

static kpse_file_format_type
find_format P2C(string, name, boolean, is_filename)
{
  kpse_file_format_type ret;
  
  if (is_filename && user_format != kpse_last_format) {
    ret = user_format;
  } else if (FILESTRCASEEQ (name, "pdftex.cfg")) {
    ret = kpse_pdftex_config_format;
  } else {
    int f;  /* kpse_file_format_type */
    unsigned name_len = strlen (name);

/* Have to rely on `try_len' being declared here, since we can't assume
   GNU C and statement expressions.  */
#define TRY_SUFFIX(ftry) (\
  try_len = (ftry) ? strlen (ftry) : 0, \
  (ftry) && try_len <= name_len \
     && FILESTRCASEEQ (ftry, name + name_len - try_len))

    f = 0;
    while (f != kpse_last_format) {
      unsigned try_len;
      const_string *ext;
      const_string ftry;
      boolean found = false;
      
      if (!kpse_format_info[f].type)
        kpse_init_format ((kpse_file_format_type)f);

      if (!is_filename) {
        /* Allow the long name, but only in the -format option.  We don't
           want a filename confused with a format name.  */
        ftry = kpse_format_info[f].type;
        found = TRY_SUFFIX (ftry);
      }
      for (ext = kpse_format_info[f].suffix; !found && ext && *ext; ext++){
        found = TRY_SUFFIX (*ext);
      }      
      for (ext = kpse_format_info[f].alt_suffix; !found && ext && *ext; ext++){
        found = TRY_SUFFIX (*ext);
      }

      if (found)
        break;

      /* Some trickery here: the extensions for kpse_fmt_format can
       * clash with other extensions in use, and we prefer for those
       * others to be preferred.  And we don't want to change the
       * integer value of kpse_fmt_format.  So skip it when first
       * enountered, then use it when we've done everything else,
       * and use it as the end-guard.
       */
      if (f == kpse_fmt_format) {
        f = kpse_last_format;
      } else if (++f == kpse_fmt_format) {
        f++;
      } else if (f == kpse_last_format) {
        f = kpse_fmt_format;
      }
    }

    /* If there was a match, f will be one past the correct value.  */
    ret = f;
  }
  
  return ret;
}



/* Return newly-allocated NULL-terminated list of strings from MATCHES
   that are prefixed with any of the subdirectories in SUBDIRS.  That
   is, for a string S in MATCHES, its dirname must end with one of the
   elements in SUBDIRS.  For instance, if subdir=foo/bar, that will
   match a string foo/bar/baz or /some/texmf/foo/bar/baz.
   
   We don't reallocate the actual strings, just the list elements.
   Perhaps later we will implement wildcards or // or something.  */

static string *
subdir_match P2C(str_list_type, subdirs,  string *, matches)
{
  string *ret = XTALLOC1 (string);
  unsigned len = 1;
  unsigned m;
  
  for (m = 0; matches[m]; m++) {
    unsigned loc;
    unsigned e;
    string s = xstrdup (matches[m]);
    for (loc = strlen (s); loc > 0 && !IS_DIR_SEP (s[loc-1]); loc--)
      ;
    while (IS_DIR_SEP (s[loc-1])) {
      loc--;
    }
    s[loc] = 0;  /* wipe out basename */
    
    for (e = 0; e < STR_LIST_LENGTH (subdirs); e++) {
      string subdir = STR_LIST_ELT (subdirs, e);
      unsigned subdir_len = strlen (subdir);
      while (subdir_len > 0 && IS_DIR_SEP (subdir[subdir_len-1])) {
        subdir_len--;
        subdir[subdir_len] = 0; /* remove trailing slashes from subdir spec */
      }
      if (FILESTRCASEEQ (subdir, s + loc - subdir_len)) {
        /* matched, save this one.  */
        ret = XRETALLOC (ret, len + 1, string);
        ret[len-1] = matches[m];
        len++;
      }
    }
    free (s);
  }
  ret[len-1] = NULL;
  return ret;
}



/* Look up a single filename NAME.  Return 0 if success, 1 if failure.  */

static unsigned
lookup P1C(string, name)
{
  int i;
  string ret = NULL;
  string *ret_list = NULL;
  
  if (user_path) {
    if (show_all) {
      ret_list = kpse_all_path_search (user_path, name);
    } else {
       ret = kpse_path_search (user_path, name, must_exist);
    }
    
  } else {
    /* No user-specified search path, check user format or guess from NAME.  */
    kpse_file_format_type fmt = find_format (name, true);

    switch (fmt) {
      case kpse_pk_format:
      case kpse_gf_format:
      case kpse_any_glyph_format:
        {
          kpse_glyph_file_type glyph_ret;
          /* Try to extract the resolution from the name.  */
          unsigned local_dpi = find_dpi (name);
          if (!local_dpi)
            local_dpi = dpi;
          ret = kpse_find_glyph (remove_suffix (name), local_dpi, fmt,
                                 &glyph_ret);
        }
        break;

      case kpse_last_format:
        /* If the suffix isn't recognized, assume it's a tex file. */
        fmt = kpse_tex_format;
        /* fall through */

      default:
        if (show_all) {
          ret_list = kpse_find_file_generic (name, fmt, must_exist, true);
        } else {
          ret = kpse_find_file (name, fmt, must_exist);
        }
    }
  }
  
  /* Turn single return into a null-terminated list for uniform treatment.  */
  if (ret) {
    ret_list = XTALLOC (2, string);
    ret_list[0] = ret;
    ret_list[1] = NULL;
  }

  /* Filter by subdirectories, if specified.  */
  if (STR_LIST_LENGTH (subdir_paths) > 0) {
    string *new_list = subdir_match (subdir_paths, ret_list);
    free (ret_list);
    ret_list = new_list;
  }

  /* Print output.  */
  if (ret_list) {
    for (i = 0; ret_list[i]; i++)
      puts (ret_list[i]);
    /* Save whether we found anything */
    ret = ret_list[0];
    free (ret_list);
  }
  
  return ret == NULL;
}

/* Reading the options.  */

#define USAGE "\
  Standalone path lookup and expansion for Kpathsea.\n\
\n\
-debug=NUM             set debugging flags.\n\
-D, -dpi=NUM           use a base resolution of NUM; default 600.\n\
-engine=STRING         set engine name to STRING.\n\
-expand-braces=STRING  output variable and brace expansion of STRING.\n\
-expand-path=STRING    output complete path expansion of STRING.\n\
-expand-var=STRING     output variable expansion of STRING.\n\
-format=NAME           use file type NAME (see list below).\n\
-help                  print this message and exit.\n\
-interactive           ask for additional filenames to look up.\n\
[-no]-mktex=FMT        disable/enable mktexFMT generation (FMT=pk/mf/tex/tfm).\n\
-mode=STRING           set device name for $MAKETEX_MODE to STRING;\n\
                       no default.\n\
-must-exist            search the disk as well as ls-R if necessary\n\
-path=STRING           search in the path STRING.\n\
-progname=STRING       set program name to STRING.\n\
-show-path=NAME        output search path for file type NAME (see list below).\n\
-var-value=STRING      output the value of variable $STRING.\n\
-version               print version number and exit.\n \
"

/* Test whether getopt found an option ``A''.
   Assumes the option index is in the variable `option_index', and the
   option table in a variable `long_options'.  */
#define ARGUMENT_IS(a) STREQ (long_options[option_index].name, a)

/* SunOS cc can't initialize automatic structs.  */
static struct option long_options[]
  = { { "D",			1, 0, 0 },
      { "all",			0, (int *) &show_all, 1 },
      { "debug",		1, 0, 0 },
      { "dpi",			1, 0, 0 },
      { "engine",		1, 0, 0 },
      { "expand-braces",	1, 0, 0 },
      { "expand-path",		1, 0, 0 },
      { "expand-var",		1, 0, 0 },
      { "format",		1, 0, 0 },
      { "help",                 0, 0, 0 },
      { "interactive",		0, (int *) &interactive, 1 },
      { "mktex",		1, 0, 0 },
      { "mode",			1, 0, 0 },
      { "must-exist",		0, (int *) &must_exist, 1 },
      { "path",			1, 0, 0 },
      { "no-mktex",		1, 0, 0 },
      { "progname",		1, 0, 0 },
      { "separator",		1, 0, 0 },
      { "subdir",		1, 0, 0 },
      { "show-path",		1, 0, 0 },
      { "var-value",		1, 0, 0 },
      { "version",              0, 0, 0 },
      { 0, 0, 0, 0 } };

static void
read_command_line P2C(int, argc,  string *, argv)
{
  int g;   /* `getopt' return code.  */
  int option_index;

  for (;;) {
    g = getopt_long_only (argc, argv, "", long_options, &option_index);

    if (g == -1)
      break;

    if (g == '?')
      exit (1);  /* Unknown option.  */

    assert (g == 0); /* We have no short option names.  */

    if (ARGUMENT_IS ("debug")) {
      kpathsea_debug |= atoi (optarg);

    } else if (ARGUMENT_IS ("dpi") || ARGUMENT_IS ("D")) {
      dpi = atoi (optarg);

    } else if (ARGUMENT_IS ("engine")) {
      engine = optarg;
      
    } else if (ARGUMENT_IS ("expand-braces")) {
      braces_to_expand = optarg;
      
    } else if (ARGUMENT_IS ("expand-path")) {
      path_to_expand = optarg;

    } else if (ARGUMENT_IS ("expand-var")) {
      var_to_expand = optarg;

    } else if (ARGUMENT_IS ("format")) {
      user_format_string = optarg;

    } else if (ARGUMENT_IS ("help")) {
      int f; /* kpse_file_format_type */
      extern KPSEDLL char *kpse_bug_address; /* from version.c */
      
      printf ("Usage: %s [OPTION]... [FILENAME]...\n", argv[0]);
      fputs (USAGE, stdout);
      putchar ('\n');
      fputs (kpse_bug_address, stdout);

      /* Have to set this for init_format to work.  */
      kpse_set_program_name (argv[0], progname);

      puts ("\nRecognized format names and their suffixes:");
      for (f = 0; f < kpse_last_format; f++) {
        const_string *ext;
        kpse_init_format ((kpse_file_format_type)f);
        printf ("%s:", kpse_format_info[f].type);
        for (ext = kpse_format_info[f].suffix; ext && *ext; ext++) {
          putchar (' ');
          fputs (*ext, stdout);
        }
        for (ext = kpse_format_info[f].alt_suffix; ext && *ext; ext++) {
          putchar (' ');
          fputs (*ext, stdout);
        }
        putchar ('\n');
      }

      exit (0);

    } else if (ARGUMENT_IS ("mktex")) {
      kpse_maketex_option (optarg, true);

    } else if (ARGUMENT_IS ("mode")) {
      mode = optarg;

    } else if (ARGUMENT_IS ("no-mktex")) {
      kpse_maketex_option (optarg, false);

    } else if (ARGUMENT_IS ("path")) {
      user_path = optarg;

    } else if (ARGUMENT_IS ("progname")) {
      progname = optarg;

    } else if (ARGUMENT_IS ("show-path")) {
      path_to_show = optarg;
      user_format_string = optarg;

    } else if (ARGUMENT_IS ("subdir")) {
      str_list_add (&subdir_paths, optarg);

    } else if (ARGUMENT_IS ("var-value")) {
      var_to_value = optarg;

    } else if (ARGUMENT_IS ("version")) {
      extern KPSEDLL char *kpathsea_version_string; /* from version.c */
      puts (kpathsea_version_string);
      puts ("Copyright 2005 Karl Berry & Olaf Weber.\n\
There is NO warranty.  You may redistribute this software\n\
under the terms of the GNU General Public License.\n\
For more information about these matters, see the files named GPL and LGPL.");
      exit (0);
    }

    /* Else it was just a flag; getopt has already done the assignment.  */
  }
  
  if (user_path && user_format_string) {
    fprintf (stderr, "-path (%s) and -format (%s) are mutually exclusive.\n",
             user_path, user_format_string);
    fputs ("Try `kpsewhich --help' for more information.\n", stderr);
    exit (1);
  }

  if (optind == argc && !var_to_expand && !braces_to_expand && !path_to_expand
                     && !path_to_show && !var_to_value) {
    fputs ("Missing argument. Try `kpsewhich --help' for more information.\n",
           stderr);
    exit (1);
  }
}

int
main P2C(int, argc,  string *, argv)
{
  unsigned unfound = 0;

  read_command_line (argc, argv);

  kpse_set_program_name (argv[0], progname);

  if (engine)
    xputenv("engine", engine);
  
  /* NULL for no fallback font.  */
  kpse_init_prog (uppercasify (kpse_program_name), dpi, mode, NULL);
  
  /* Have to do this after setting the program name.  */
  if (user_format_string)
    user_format = find_format (user_format_string, false);
  
  /* Variable expansion.  */
  if (var_to_expand)
    puts (kpse_var_expand (var_to_expand));

  /* Brace expansion. */
  if (braces_to_expand)
    puts (kpse_brace_expand (braces_to_expand));
  
  /* Path expansion. */
  if (path_to_expand)
    puts (kpse_path_expand (path_to_expand));

  /* Show a search path. */
  if (path_to_show) {
    if (user_format != kpse_last_format) {
      if (!kpse_format_info[user_format].type) /* needed if arg was numeric */
        kpse_init_format (user_format);
      puts (kpse_format_info[user_format].path);
    } else {
      WARNING1 ("kpsewhich: Ignoring unknown file type `%s'", path_to_show);
    }
  }

  /* Var to value. */
  if (var_to_value) {
    string value = kpse_var_value (var_to_value);
    if (!value) {
      unfound++;
      value="";
    }
    puts (value);
  }
  
  /* --subdir must imply --all, since we filter here after doing the
     search, rather than inside the search itself.  */
  if (STR_LIST_LENGTH (subdir_paths) > 0) {
    show_all = 1;
  }
  
  for (; optind < argc; optind++) {
    unfound += lookup (argv[optind]);
  }

  if (interactive) {
    for (;;) {
      string name = read_line (stdin);
      if (!name || STREQ (name, "q") || STREQ (name, "quit")) break;
      unfound += lookup (name);
      free (name);
    }
  }
  
  return unfound > 255 ? 1 : unfound;
}
